### ISSUE TITLE
downsample_traingenes.pl / GeneMark-ET error

### ISSUE THREAD

**Issue by ricardo-aaron** [Gaius-Augustus/BRAKER#251]
Hi,

I'm getting this error:
ERROR in file /path_to/BRAKER/scripts/braker.pl at line 6767
Failed to execute: perl /path_to/BRAKER/scripts/downsample_traingenes.pl --in_gtf=/path_to/outdir/GeneMark-ET/genemark.f.good.gtf --out_gtf=/path_to/outdir/GeneMark-ET/genemark.d.gtf --lambda=2 1> /path_to/outdir/downsample_traingenes.log 2> /path_to/outdir/errors/downsample_traingenes.err

downsample_traingenes.err has:
Could not open file /path_to/outdir/GeneMark-ET/genemark.f.good.gtf!

the GeneMark-ET dir contents are:
total 92603
drwxrwxr-x 3 rchavez rchavez        7 Aug 12 14:06 data
-rw-rw-r-- 1 rchavez rchavez        7 Aug 12 14:17 genemark.average_gene_length.out
-rw-rw-r-- 1 rchavez rchavez 46080887 Aug 12 14:17 genemark.c.gtf
-rw-rw-r-- 1 rchavez rchavez 46080887 Aug 12 14:17 genemark.gtf
-rw-rw-r-- 1 rchavez rchavez   309541 Aug 12 14:17 gmes.log
-rw-rw-r-- 1 rchavez rchavez  1988427 Aug 12 14:15 gmhmm.mod
drwxrwxr-x 2 rchavez rchavez       10 Aug 12 14:15 info
drwxrwxr-x 4 rchavez rchavez        5 Aug 12 14:15 output
drwxrwxr-x 9 rchavez rchavez       13 Aug 12 14:15 run
-rw-rw-r-- 1 rchavez rchavez    17192 Aug 12 14:06 run.cfg

this with BRAKER-2.5.1

**Comment by tomasbruna** [comment]
Hello @ricardo-aaron,

did you try to run the `test1.sh` example in the [example folder](https://github.com/Gaius-Augustus/BRAKER/tree/master/example/tests) and are you getting the same error with the example input? More details on running the tests are in the [example section of the readme file](https://github.com/Gaius-Augustus/BRAKER#example-data).

If you have not already done so, please also pull/clone the latest version of BRAKER from the master branch.

Best,
Tomas

**Comment by ricardo-aaron** [comment]
just cloned, same error:
# Wed Aug 12 17:46:46 2020: Log information is stored in file /pool/software/BRAKER/example/tests/test1/braker.log
ERROR in file /pool/software/BRAKER/scripts/braker.pl at line 6689
Failed to execute: /usr/bin/python3 /pool/software/BRAKER/scripts/ensure_n_training_genes.py --goodGenes /pool/software/BRAKER/example/tests/test1/GeneMark-ET/genemark.f.good.gtf --badGenes /pool/software/BRAKER/example/tests/test1/GeneMark-ET/genemark.f.bad.gtf --N 4000 1>/pool/software/BRAKER/example/tests/test1/ensure_min_n_training_genes.stdout 2>/pool/software/BRAKER/example/tests/test1/errors/ensure_min_n_training_genes.stderr

contents of ensure_min_n_training_genes.stderr:
Traceback (most recent call last):
  File "/pool/software/BRAKER/scripts/ensure_n_training_genes.py", line 135, in <module>
    main()
  File "/pool/software/BRAKER/scripts/ensure_n_training_genes.py", line 101, in main
    goodGenes = loadGenes(args.goodGenes)
  File "/pool/software/BRAKER/scripts/ensure_n_training_genes.py", line 62, in loadGenes
    for row in csv.reader(open(genesFile), delimiter='\t'):
FileNotFoundError: [Errno 2] No such file or directory: '/pool/software/BRAKER/example/tests/test1/GeneMark-ET/genemark.f.good.gtf'

**Comment by tomasbruna** [comment]
It's a good thing the error is reproducible with the test file. There seems to be something wrong with the `filterGenemark.pl` script, can you share the full `/pool/software/BRAKER/example/tests/test1/braker.log`? Thanks!

**Comment by ricardo-aaron** [comment]
braker.log of test1:

#**********************************************************************************
#                               BRAKER CONFIGURATION                               
#**********************************************************************************
# BRAKER CALL: /elpool/software/BRAKER/scripts/braker.pl --genome=../genome.fa --bam=../RNAseq.bam --softmasking --workingdir=test1 --cores 8 --gm_max_intergenic 10000
# Wed Aug 12 18:30:55 2020: braker.pl version 2.1.5
# Wed Aug 12 18:30:55 2020: Creating directory /elpool/software/BRAKER/example/tests/test1.
# Wed Aug 12 18:30:55 2020: Configuring of BRAKER for using external tools...
# Wed Aug 12 18:30:55 2020: Found environment variable $AUGUSTUS_CONFIG_PATH. Setting $AUGUSTUS_CONFIG_PATH to /elpool/software/Augustus/config
# Wed Aug 12 18:30:55 2020: Found environment variable $AUGUSTUS_BIN_PATH. Setting $AUGUSTUS_BIN_PATH to /elpool/software/Augustus/bin
# Wed Aug 12 18:30:55 2020: Found environment variable $AUGUSTUS_SCRIPTS_PATH. Setting $AUGUSTUS_SCRIPTS_PATH to /elpool/software/Augustus/scripts
# Wed Aug 12 18:30:55 2020: Did not find environment variable $PYTHON3_PATH
# Wed Aug 12 18:30:55 2020: Trying to guess $PYTHON3_PATH from location of python3 executable that is available in your $PATH
# Wed Aug 12 18:30:55 2020: Setting $PYTHON3_PATH to /usr/bin
# Wed Aug 12 18:30:55 2020: Found environment variable $GENEMARK_PATH. Setting $GENEMARK_PATH to /elpool/software/gmes_linux_64
# Wed Aug 12 18:30:55 2020: Did not find environment variable $BAMTOOLS_PATH (either variable does not exist, or the path given in variable does not exist). Will try to set this variable in a different way, later.
# Wed Aug 12 18:30:55 2020: Trying to guess $BAMTOOLS_BIN_PATH from location of bamtools executable that is available in your $PATH
# Wed Aug 12 18:30:55 2020: Setting $BAMTOOLS_BIN_PATH to /elpool/software/bamtools/bin
# Wed Aug 12 18:30:55 2020: Did not find environment variable $SAMTOOLS_PATH  (either variable does not exist, or the path given in variable doesnot exist). Will try to set this variable in a different way, later.
# Wed Aug 12 18:30:55 2020: Trying to guess $SAMTOOLS_PATH from location of samtools executable in your $PATH
# Wed Aug 12 18:30:55 2020: Setting $SAMTOOLS_PATH to /elpool/software/samtools/bin
# Wed Aug 12 18:30:55 2020: Did not find environment variable $DIAMOND_PATH
# Wed Aug 12 18:30:55 2020: Trying to guess $DIAMOND_PATH from location of diamond executable that is available in your $PATH
# Wed Aug 12 18:30:55 2020: Setting $DIAMOND_PATH to /elpool/software/diamond/bin
# Wed Aug 12 18:30:55 2020: Did not find environment variable $CDBTOOLS_PATH
# Wed Aug 12 18:30:55 2020: Trying to guess $CDBTOOLS_PATH from location of cdbfasta executable that is available in your $PATH
# Wed Aug 12 18:30:55 2020: Setting $CDBTOOLS_PATH to /elpool/software/cdbfasta
#*********
# IMPORTANT INFORMATION: no species for identifying the AUGUSTUS  parameter set that will arise from this BRAKER run was set. BRAKER will create an AUGUSTUS parameter set with name Sp_7. This parameter set can be used for future BRAKER/AUGUSTUS prediction runs for the same species. It is usually not necessary to retrain AUGUSTUS with novel extrinsic data if a high quality parameter set already exists.
#*********
#**********************************************************************************
#                               CREATING DIRECTORY STRUCTURE                       
#**********************************************************************************
# Wed Aug 12 18:30:55 2020: creating file that contains citations for this BRAKER run at /elpool/software/BRAKER/example/tests/test1/what-to-cite.txt...
# Wed Aug 12 18:30:55 2020: create working directory /elpool/software/BRAKER/example/tests/test1/GeneMark-ET.
mkdir /elpool/software/BRAKER/example/tests/test1/GeneMark-ET
# Wed Aug 12 18:30:55 2020: create working directory /elpool/software/BRAKER/example/tests/test1/species
mkdir /elpool/software/BRAKER/example/tests/test1/species
# Wed Aug 12 18:30:55 2020: create working directory /elpool/software/BRAKER/example/tests/test1/errors
mkdir /elpool/software/BRAKER/example/tests/test1/errors
# Wed Aug 12 18:30:55 2020: changing into working directory /elpool/software/BRAKER/example/tests/test1
cd /elpool/software/BRAKER/example/tests/test1
# Wed Aug 12 18:30:55 2020: Creating parameter template files for AUGUSTUS with new_species.pl
# Wed Aug 12 18:30:55 2020: new_species.pl will create parameter files for species Sp_7 in /elpool/software/Augustus/config/species/Sp_7
perl /elpool/software/Augustus-3.3.3/scripts/new_species.pl --species=Sp_7 --AUGUSTUS_CONFIG_PATH=/elpool/software/Augustus/config 1> /dev/null 2>/elpool/software/BRAKER/example/tests/test1/errors/new_species.stderr
# Wed Aug 12 18:30:55 2020: check_fasta_headers(): Checking fasta headers of file /elpool/software/BRAKER/example/tests/../genome.fa
#**********************************************************************************
#                               PROCESSING HINTS                                   
#**********************************************************************************
# Wed Aug 12 18:30:55 2020: Converting bam files to hints
# Wed Aug 12 18:30:55 2020: Checking bam headers
# Wed Aug 12 18:30:55 2020: create header file /elpool/software/BRAKER/example/tests/test1/RNAseq_header.sam
/elpool/software/bamtools/bin/bamtools header -in /elpool/software/BRAKER/example/tests/../RNAseq.bam > /elpool/software/BRAKER/example/tests/test1/RNAseq_header.sam
# Wed Aug 12 18:30:55 2020: Deleting SAM header file /elpool/software/BRAKER/example/tests/test1/RNAseq_header.sam (will not be needed from here on)
# Wed Aug 12 18:30:55 2020: Deleting /elpool/software/BRAKER/example/tests/test1/RNAseq_new_header.sam
# Wed Aug 12 18:30:55 2020: make hints from BAM file /elpool/software/BRAKER/example/tests/../RNAseq.bam
/elpool/software/Augustus/bin/bam2hints --intronsonly --in=/elpool/software/BRAKER/example/tests/../RNAseq.bam --out=/elpool/software/BRAKER/example/tests/test1/bam2hints.temp.0.gff 1> /elpool/software/BRAKER/example/tests/test1/errors/bam2hints.0.stdout 2>/elpool/software/BRAKER/example/tests/test1/errors/bam2hints.0.stderr
# Wed Aug 12 18:31:12 2020: add hints from BAM file /elpool/software/BRAKER/example/tests/../RNAseq.bam to hints file
cat /elpool/software/BRAKER/example/tests/test1/bam2hints.temp.0.gff >>/elpool/software/BRAKER/example/tests/test1/hintsfile.temp.gff
# Wed Aug 12 18:31:12 2020: Checking for hints of src=C and with grp tags that should not be joined according to multiplicity
# Wed Aug 12 18:31:12 2020: Joining hints that are identical (& from the same source) into multiplicity hints (input file /elpool/software/BRAKER/example/tests/test1/tmp_merge_hints.gff)
# Wed Aug 12 18:31:12 2020: sort hints of type rnaseq
cat /elpool/software/BRAKER/example/tests/test1/tmp_merge_hints.gff | sort -n -k 4,4 | sort -s -n -k 5,5 | sort -s -n -k 3,3 | sort -s -k 1,1 >/elpool/software/BRAKER/example/tests/test1/hints.rnaseq.temp.sort.gff
# Wed Aug 12 18:31:12 2020: join multiple hints
perl /elpool/software/Augustus-3.3.3/scripts/join_mult_hints.pl </elpool/software/BRAKER/example/tests/test1/hints.rnaseq.temp.sort.gff >/elpool/software/BRAKER/example/tests/test1/tmp_merge_hints.gff 2>/elpool/software/BRAKER/example/tests/test1/errors/join_mult_hints.rnaseq.stderr
mv /elpool/software/BRAKER/example/tests/test1/tmp_merge_hints.gff /elpool/software/BRAKER/example/tests/test1/hintsfile.temp.gff
# Wed Aug 12 18:31:12 2020: filter introns, find strand and change score to 'mult' entry
perl /elpool/software/Augustus-3.3.3/scripts/filterIntronsFindStrand.pl /elpool/software/BRAKER/example/tests/test1/genome.fa /elpool/software/BRAKER/example/tests/test1/hintsfile.temp.gff --score 1>>/elpool/software/BRAKER/example/tests/test1/hintsfile.gff 2>/elpool/software/BRAKER/example/tests/test1/errors/filterIntronsFindStrand.stderr
# Wed Aug 12 18:31:12 2020: rm /elpool/software/BRAKER/example/tests/test1/hintsfile.temp.gff
# Wed Aug 12 18:31:12 2020: Preparing hints for running GeneMark
# Wed Aug 12 18:31:12 2020: Filtering intron hints for GeneMark from /elpool/software/BRAKER/example/tests/test1/hintsfile.gff...
cat /elpool/software/BRAKER/example/tests/test1/genemark_hintsfile.gff.rnaseq | sort -n -k 4,4 | sort -s -n -k 5,5 | sort -s -k 3,3 | sort -s -k 1,1 | /elpool/software/Augustus-3.3.3/scripts/join_mult_hints.pl > /elpool/software/BRAKER/example/tests/test1/genemark_hintsfile.gff.rnaseq.tmp
mv /elpool/software/BRAKER/example/tests/test1/genemark_hintsfile.gff.rnaseq.tmp /elpool/software/BRAKER/example/tests/test1/genemark_hintsfile.gff
#**********************************************************************************
#                              RUNNING GENEMARK-EX                                 
#**********************************************************************************
# Wed Aug 12 18:31:12 2020: Checking whether file /elpool/software/BRAKER/example/tests/test1/genemark_hintsfile.gff contains enough hints and sufficient multiplicity information...
# Wed Aug 12 18:31:12 2020: Executing GeneMark-ET
# Wed Aug 12 18:31:12 2020: changing into GeneMark-ET directory /elpool/software/BRAKER/example/tests/test1/GeneMark-ET
cd /elpool/software/BRAKER/example/tests/test1/GeneMark-ET
# Wed Aug 12 18:31:12 2020: Executing gmes_petap.pl
perl /elpool/software/gmes_linux_64/gmes_petap.pl --verbose --sequence=/elpool/software/BRAKER/example/tests/test1/genome.fa --ET=/elpool/software/BRAKER/example/tests/test1/genemark_hintsfile.gff --cores=8 --gc_donor 0.001 --soft_mask auto --max_intergenic=10000 1>/elpool/software/BRAKER/example/tests/test1/GeneMark-ET.stdout 2>/elpool/software/BRAKER/example/tests/test1/errors/GeneMark-ET.stderr
# Wed Aug 12 18:32:14 2020: change to working directory /elpool/software/BRAKER/example/tests/test1
cd /elpool/software/BRAKER/example/tests/test1
# Wed Aug 12 18:32:14 2020: Filtering output of GeneMark for generating training genes for AUGUSTUS
# Wed Aug 12 18:32:14 2020: Checking whether hintsfile contains single exon CDSpart hints or start/stop hints
# Wed Aug 12 18:32:14 2020: filtering GeneMark genes by intron hints
perl /elpool/software/Augustus-3.3.3/scripts/filterGenemark.pl --genemark=/elpool/software/BRAKER/example/tests/test1/GeneMark-ET/genemark.gtf --hints=/elpool/software/BRAKER/example/tests/test1/hintsfile.gff --randomSeed=1 1>/elpool/software/BRAKER/example/tests/test1/filterGenemark.stdout 2>/elpool/software/BRAKER/example/tests/test1/errors/filterGenemark.stderr
# Wed Aug 12 18:32:14 2020: Ensuring at least 4000 genes in training file 
/usr/bin/python3 /elpool/software/BRAKER/scripts/ensure_n_training_genes.py --goodGenes /elpool/software/BRAKER/example/tests/test1/GeneMark-ET/genemark.f.good.gtf --badGenes /elpool/software/BRAKER/example/tests/test1/GeneMark-ET/genemark.f.bad.gtf --N 4000 1>/elpool/software/BRAKER/example/tests/test1/ensure_min_n_training_genes.stdout 2>/elpool/software/BRAKER/example/tests/test1/errors/ensure_min_n_training_genes.stderr

**Comment by tomasbruna** [comment]
The problem originates from this system call:

    perl /elpool/software/Augustus-3.3.3/scripts/filterGenemark.pl --genemark=/elpool/software/BRAKER/example/tests/test1/GeneMark-ET/genemark.gtf --hints=/elpool/software/BRAKER/example/tests/test1/hintsfile.gff --randomSeed=1 1>/elpool/software/BRAKER/example/tests/test1/filterGenemark.stdout 2>/elpool/software/BRAKER/example/tests/test1/errors/filterGenemark.stderr

I do not know how it is possible that BRAKER found the `filterGenemark.pl` script in `/elpool/software/Augustus-3.3.3/scripts/`, it is not a part of AUGUSTUS - see https://github.com/Gaius-Augustus/Augustus/tree/master/scripts. Somehow, this unexpected script is not working properly.

Can you check why this script is there and delete it? Once it is gone, BRAKER should use `filterGenemark.pl` provided in the [BRAKER scripts](https://github.com/Gaius-Augustus/BRAKER/tree/master/scripts) folder and the problem should disappear.

Best,
Tomas

**Comment by ricardo-aaron** [comment]
Now that you mention it I think I copied the script there because BRAKER-2.1.5.tar.gz wasn't finding it... But I deleted that BRAKER, cloned, re-sourced my .bashrc, and "which filterGenemark.pl" was answering "/elpool/software/BRAKER/scripts/filterGenemark.pl"... In any case, I deleted Augustus-3-3-3, git cloned Augustus, and test1.sh (and test2.sh) success!
Thank you Thomas, you're the best. It's amazing how you guys can quickly find the error among a bunch of lines in log files.
